pkg/config/types.go

```diff
@@ -68,6 +68,21 @@ type UserMapping struct {
 	Groups []string `json:"groups"`
 }
 
+// ARNLikeMapping is a dynamic mapping of a matching AWS IAM ARNs to a
+// Kubernetes username and a list of Kubernetes groups
+type ARNLikeMapping struct {
+	// ARNLike is a string used to match AWS ARNs with globs. (e.g., "arn:aws:iam::000000000000:role/Kubernetes*")
+	ARNLike string `json:"arnLike"`
+
+	// Username is the username pattern that principals matching this
+	// string will have in Kubernetes.
+	Username string `json:"username"`
+
+	// Groups is a list of Kubernetes groups principals matching this string will
+	// authenticate as (e.g., `system:masters`)
+	Groups []string `json:"groups"`
+}
```

Comment:
After reading through this PR a number of times, it struck me that it might be better to add a `UserMapping.UserARNLike` and a `RoleMapping.RoleARNLike` field instead of adding in a new `ARNLikeMapping` struct to the configuration. This would still be backwards-compatible with older configurations and more neatly nest the ARNLike functionality into the existing configuration blocks (for `users` and `roles` respectively)

It would also mean that you would not need to modify function signatures and return values for things like `add()` and `ParseMap()`

Reply:
Ok I'll give that a go. I wonder if it would be better as a plural (i.e. `UserARNLikes []string`) but on the other hand it's probably a little tidier and I think it still meets one of my use-cases - where I have many similarly named IAM roles I'd like to map to a static series of groups